File: bindings/aitp-py/tests/test_manifest_verify.py
"""`verify_manifest_json` free function — Python SDK."""

import json

import pytest

import aitp


def _signed_manifest():
    a = aitp.AitpAgent.generate()
    return a.build_manifest(
        display_name="enrollee",
        handshake_endpoint="http://localhost:9000/aitp/handshake/",
        offered_caps=["demo.write"],
    )


def test_verify_manifest_accepts_freshly_built():
    aitp.verify_manifest_json(_signed_manifest())


def test_verify_manifest_rejects_tampered_payload():
    env = json.loads(_signed_manifest())
    # Mutate a load-bearing field — display_name is part of the JCS body.
    env["manifest"]["display_name"] = "imposter"
    tampered = json.dumps(env)
    with pytest.raises(RuntimeError, match="manifest verification failed"):
        aitp.verify_manifest_json(tampered)


def test_verify_manifest_rejects_garbage():
    with pytest.raises(ValueError, match="invalid manifest JSON"):
        aitp.verify_manifest_json("not json")


# ── Typed failure causes ─────────────────────────────────────────────────
#
# Added in 0.6.1. Until then `verify_manifest_json` raised a bare RuntimeError
# while `verify_revocation_list` carried a stable `.code` — so a caller who
# wanted to tell "this manifest expired" from "this manifest is forged" had to
# substring-match prose on one side and could branch on a code on the other.
# That asymmetry is the same shape as the missing-binding defect this family
# spent a release on: one artifact has the surface, its sibling does not.


def _code_for(envelope_json: str) -> str:
    try:
        aitp.verify_manifest_json(envelope_json)
    except aitp.ManifestVerificationError as exc:
        return exc.code
    raise AssertionError("expected verification to fail, but it succeeded")


def test_every_failure_cause_is_recoverable_without_parsing_a_message():
    """`.code` is the contract; the message wording is not."""
    import json

    agent = aitp.AitpAgent.generate()
    good = agent.build_manifest(
        display_name="peer",
        handshake_endpoint="http://localhost:9/aitp/handshake/",
        offered_caps=["demo.x"],
    )
    assert aitp.verify_manifest_json(good) is None

    tampered = json.loads(good)
    tampered["manifest"]["display_name"] = "evil"
    assert _code_for(json.dumps(tampered)) == "signature_invalid"

    # A negative TTL back-dates expires_at before published_at, so the
    # manifest is expired the instant it is minted — deterministic, no sleep.
    expired = agent.build_manifest(
        display_name="peer",
        handshake_endpoint="http://localhost:9/aitp/handshake/",
        offered_caps=["demo.x"],
        ttl_secs=-1,
    )
    assert _code_for(expired) == "expired"


def test_expired_is_distinct_from_signature_invalid():
    """The distinction the playground previously had to get by substring match.

    "That peer's manifest went stale" and "someone tampered with it" send an
    operator to different places; a caller that cannot separate them cannot
    alert on the second without drowning in the first.
    """
    import json

    agent = aitp.AitpAgent.generate()
    stale = agent.build_manifest(
        display_name="peer",
        handshake_endpoint="http://localhost:9/aitp/handshake/",
        offered_caps=["demo.x"],
        ttl_secs=-1,
    )
    forged = json.loads(
        agent.build_manifest(
            display_name="peer",
            handshake_endpoint="http://localhost:9/aitp/handshake/",
            offered_caps=["demo.x"],
        )
    )
    forged["manifest"]["aid"] = aitp.AitpAgent.generate().aid

    assert _code_for(stale) == "expired"
    assert _code_for(json.dumps(forged)) == "signature_invalid"

File: bindings/aitp-node/tests/test_manifest_verify.mjs
// `verifyManifestJson` free function — Node SDK.
//
// Mirrors bindings/aitp-py/tests/test_manifest_verify.py so the two
// SDKs stay at parity on the control-plane manifest-enrollment path.
//
// Build first:  npm run build:debug
// Then run:     node --test tests/

import test from 'node:test';
import assert from 'node:assert/strict';

import { AitpAgent, verifyManifestJson } from '../index.js';

function signedManifest() {
  const a = AitpAgent.generate();
  return a.buildManifest({
    displayName: 'enrollee',
    handshakeEndpoint: 'http://localhost:9000/aitp/handshake/',
    offeredCaps: ['demo.write'],
  });
}

test('verifyManifestJson accepts a freshly built manifest', () => {
  // Returns void on success; throws on failure.
  verifyManifestJson(signedManifest());
});

test('verifyManifestJson rejects a tampered payload', () => {
  const env = JSON.parse(signedManifest());
  // display_name is part of the signed JCS body — mutating it breaks
  // the signature.
  env.manifest.display_name = 'imposter';
  assert.throws(() => verifyManifestJson(JSON.stringify(env)));
});

test('verifyManifestJson rejects garbage input', () => {
  assert.throws(() => verifyManifestJson('not json'));
});

// ── Typed failure causes ─────────────────────────────────────────────────
//
// Added in 0.6.1. Until then verifyManifestJson threw a generic error while
// verifyRevocationList carried a stable error.code — so a caller wanting to
// tell "this manifest expired" from "this manifest is forged" had to parse
// prose on one side and could branch on a code on the other. Same shape as
// the missing-binding defect this family spent a release on: one artifact has
// the surface, its sibling does not.

function causeOf(fn) {
  try {
    fn();
  } catch (err) {
    // err.code is the contract. Deliberately NOT parsed out of err.message:
    // a caller string-matching a message pins program output as an expected
    // value, which is the bug class this binding exists to remove.
    return err.code;
  }
  throw new Error('expected verification to fail, but it succeeded');
}

test('every manifest failure cause is recoverable without parsing prose', () => {
  const a = AitpAgent.generate();
  const mk = (extra = {}) =>
    a.buildManifest({
      displayName: 'peer',
      handshakeEndpoint: 'http://localhost:9000/aitp/handshake/',
      offeredCaps: ['demo.write'],
      ...extra,
    });

  assert.equal(verifyManifestJson(mk()), undefined);

  const tampered = JSON.parse(mk());
  tampered.manifest.display_name = 'imposter';
  assert.equal(causeOf(() => verifyManifestJson(JSON.stringify(tampered))), 'signature_invalid');

  // A negative TTL back-dates expires_at before published_at, so the manifest
  // is expired the instant it is minted — deterministic, and no sleep.
  assert.equal(causeOf(() => verifyManifestJson(mk({ ttlSecs: -1 }))), 'expired');

  assert.equal(causeOf(() => verifyManifestJson('not json at all')), 'malformed');
});

// ── nowUnixSecs override ─────────────────────────────────────────────────
//
// verifyRevocationList has always accepted a clock override so tests can
// pin "expired" deterministically instead of faking the system clock or
// minting an already-negative TTL. verifyManifestJson lacked the same
// parameter (aitp-rs#102) — a caller could not distinguish "this manifest
// lapsed" from "this manifest is forged" without either faking system time
// or, as the tests above do, minting a manifest that's born expired. This
// exercises the override on an otherwise-valid (positive-TTL) manifest, so
// it proves the parameter actually reaches the verifier rather than just
// being accepted and ignored.

test('nowUnixSecs overrides the verification clock to accept a normally-valid manifest', () => {
  const manifest = signedManifest();
  const soon = Math.floor(Date.now() / 1000) + 60;
  assert.equal(verifyManifestJson(manifest, soon), undefined);
});

test('nowUnixSecs overrides the verification clock to expire a normally-valid manifest', () => {
  const manifest = signedManifest();
  const farFuture = Math.floor(Date.now() / 1000) + 10_000_000;
  assert.equal(causeOf(() => verifyManifestJson(manifest, farFuture)), 'expired');
});

test('expired is distinct from signature_invalid', () => {
  // "That peer's manifest went stale" and "someone tampered with it" send an
  // operator to different places; a caller that cannot separate them cannot
  // alert on the second without drowning in the first.
  const a = AitpAgent.generate();
  const mk = (extra = {}) =>
    a.buildManifest({
      displayName: 'peer',
      handshakeEndpoint: 'http://localhost:9000/aitp/handshake/',
      offeredCaps: ['demo.write'],
      ...extra,
    });
  const forged = JSON.parse(mk());
  forged.manifest.aid = AitpAgent.generate().aid;

  assert.notEqual(
    causeOf(() => verifyManifestJson(mk({ ttlSecs: -1 }))),
    causeOf(() => verifyManifestJson(JSON.stringify(forged))),
  );
});
